### PR TITLE
[Merged by Bors] - feat: port Algebra.Module.LocalizedModule

### PR DESCRIPTION
**Pull request by yuma-mizuno**
- [x] depends on: #3414 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 1]
```diff
@@ -0,0 +1,1122 @@
...
+theorem mk'_surjective : Function.Surjective (Function.uncurry <| mk' f : M × S → M') := by
+  intro x
+  obtain ⟨⟨m, s⟩, e : s • x = f m⟩ := IsLocalizedModule.surj S f x
+  exact ⟨⟨m, s⟩, mk'_eq_iff.mpr e.symm⟩
+#align is_localized_module.mk'_surjective IsLocalizedModule.mk'_surjective
+
+section Algebra
+
+-- Porting note: Manually added since `inferInstance` fails.
+@[nolint docBlame]
```
Comment (semorrison) [inline Mathlib/Algebra/Module/LocalizedModule.lean:1087]: Please don't add `@[nolint docBlame]`.

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 2]
```diff
@@ -0,0 +1,1122 @@
...
+instance {M : Type _} [AddCommGroup M] [Module R M] : AddCommGroup (LocalizedModule S M) :=
+  { show AddCommMonoid (LocalizedModule S M) by infer_instance,
+    show Neg (LocalizedModule S M) by infer_instance with
+    add_left_neg := add_left_neg' }
+
+theorem mk_neg {M : Type _} [AddCommGroup M] [Module R M] {m : M} {s : S} : mk (-m) s = -mk m s :=
+  rfl
+#align localized_module.mk_neg LocalizedModule.mk_neg
+
+set_option maxHeartbeats 400000 in
```
Comment (semorrison) [inline Mathlib/Algebra/Module/LocalizedModule.lean:255]: Can you try this post #3414? If that fixes it, please make this PR dependent on that.

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 3]
```diff
@@ -0,0 +1,1122 @@
...
+  induction' x using LocalizedModule.induction_on with mx sx
+  induction' y using LocalizedModule.induction_on with my sy
+  induction' z using LocalizedModule.induction_on with mz sz
+  simp only [mk_add_mk, smul_add]
+  refine' mk_eq.mpr ⟨1, _⟩
+  rw [one_smul, one_smul]
+  congr 1
+  · rw [mul_assoc]
+  · rw [eq_comm, mul_comm, add_assoc, mul_smul, mul_smul, ← mul_smul sx sz, mul_comm, mul_smul]
+-- #align localized_module.add_assoc' localized_module.add_assoc'
```
Comment (semorrison) [inline Mathlib/Algebra/Module/LocalizedModule.lean:179]: Why are these `#align`'s commented out? Perhaps leave a comment explaining.

(If the reason is that you made something `private`, either don't make it private (which is usually a bad idea) or explain why.)

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 4]
```diff
@@ -0,0 +1,1122 @@
...
+              -- Put everything in the same shape, sorting the terms using `simp`
+              have hu1' := congr_arg ((· • ·) (u2 * s2 * s2')) hu1
+              have hu2' := congr_arg ((· • ·) (u1 * s1 * s1')) hu2
+              simp only [smul_add, ← mul_smul, smul_assoc, mul_assoc, mul_comm,
+                mul_left_comm] at hu1' hu2' ⊢
+              rw [hu1', hu2']⟩
+
+theorem mk_add_mk {m1 m2 : M} {s1 s2 : S} :
+    mk m1 s1 + mk m2 s2 = mk (s2 • m1 + s1 • m2) (s1 * s2) :=
+-- Porting note: Simplified thanks to [linter.unreachableTactic].
```
Comment (semorrison) [inline Mathlib/Algebra/Module/LocalizedModule.lean:165]:
```suggestion
```
No need to record simple fixes like this that happened during the port, I think. We'll never do anything with this porting note except delete it!

File: Mathlib/Algebra/Module/LocalizedModule.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Algebra/Module/LocalizedModule.lean:1087]: Perhaps this one is also worth investigating with #3414?

Comment (yuma-mizuno) [reply Mathlib/Algebra/Module/LocalizedModule.lean]: As you guessed, merging #3414 fixed it!

File: Mathlib/Algebra/Module/LocalizedModule.lean (hunk 2 above)
Comment (yuma-mizuno) [reply Mathlib/Algebra/Module/LocalizedModule.lean:240]: This seem to be needed even after merging #3414.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3414
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3414~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**